### DISCUSSION TITLE
Parameter to Force TCP for RTSP streams

### DISCUSSION THREAD

**Discussion by Wolfleader101** [CommunityToolkit/Maui#1715] [New Feature Discussions]
Hi,

I am proposing the addition of a `ForceTCP` property of some kind on the `MediaElement` component.

This proposal stems from an issue I have had with ExoPlayer on Android, where it tries to use UDP first. However, this causes issues when connecting to some IP Camera RTSP streams.

A related issue here: https://github.com/androidx/media/issues/357

As you can see from the issue, a workaround for the issue for ExoPlayer is
```java
// Create an RTSP media source pointing to an RTSP uri.
MediaSource mediaSource =
    new RtspMediaSource.Factory()
        .setForceUseRtpTcp(true)
        .createMediaSource(MediaItem.fromUri(rtspUri));
// Create a player instance.
ExoPlayer player = new ExoPlayer.Builder(context).build();
// Set the media source to be played.
player.setMediaSource(mediaSource);
// Prepare the player.
player.prepare();
```

On the MAUI side, it would be nice for this to be a property.
i.e
### XAML Usage

```xaml
            <toolkit:MediaElement x:Name="mediaElement"
                                  ShouldAutoPlay="True"
                                  ShouldShowPlaybackControls="True"
                                  Source="rtsp://cameraip:554/cam/1"
                                  HeightRequest="300"
                                  WidthRequest="400"
                                  ForceTCP="True" />
```

### C# Usage

```cs

mediaElement.ForceTCP = true; // false by default
```

I look forward to a reply!

**Comment by Wolfleader101** [comment]
I have created a fork with this feature implemented here
https://github.com/Wolfleader101/Maui/tree/should-force-rtp-tcp-proposal


Specifically, the commit:
https://github.com/Wolfleader101/Maui/commit/7722f9c5af6e0d44be2900e0eee3d6449b8d45f8

I can also confirm that this does indeed work.

Usage follows naming convention of other boolean properties
```xaml
            <toolkit:MediaElement x:Name="mediaElement"
                                  ShouldAutoPlay="True"
                                  ShouldShowPlaybackControls="False"
                                  ShouldForceRtpTcp="True"
                                  Source=""
                                  HeightRequest="300"
                                  WidthRequest="400"
                       />
```

**Reply by pictos** [reply]
Very nice that you move this forward. But how that woul be in other platforms?

**Reply by Wolfleader101** [reply]
I am looking into ios implementation now, although I am not sure if they have this issue as it's a very niche issue related to the way Exoplayer initiates its RTSP connections.

Its also worth noting that RTSP doesn't seem to work on windows, as I don't believe it's supported by Windows Media Player.
I am not even sure if it is supported by ios ( I don't have a mac to test with)

**Reply by pictos** [reply]
If you want to implement it just for android, that can be possible using the Platform Specific API. But if we can implement it on all platforms will be better

**Reply by Wolfleader101** [reply]
> If you want to implement it just for android, that can be possible using the Platform Specific API. But if we can implement it on all platforms will be better

After further digging, it's not actually possible on any other platform as it is only androids exoplayer that supports RTSP

**Reply by bijington** [reply]
We are hopeful that we can migrate away from Exoplayer soon so we should reevaluate this once that work is done

**Comment by ghost** [comment]
[Cross-platform .NET/Mono bindings](https://github.com/videolan/libvlcsharp/tree/3.x) for LibVLC has [LGPL-2.1 license](https://github.com/videolan/libvlcsharp/tree/3.x#LGPL-2.1-1-ov-file) + available commercial licence. It does work on iOS and Android - MAUI, [however the licence is not MIT](https://www.tldrlegal.com/license/gnu-lesser-general-public-license-v2-1-lgpl-2-1).

On the iOs it take some time for video to appear (20+ seconds).
On Android it is almost instant.

**Code example by [CoPilot AI](https://copilot.microsoft.com/):**
```
*.csproject
                <PackageReference Include="LibVLCSharp.MAUI" Version="3.9.3" />
		<PackageReference Include="VideoLAN.LibVLC.Android" Version="3.5.3" />
		<PackageReference Include="VideoLAN.LibVLC.iOS" Version="3.3.18" />

MauiProgram.cs
 MauiAppBuilder builder = MauiApp.CreateBuilder();
                builder
                    .UseMauiApp<App>()
                    .UseLibVLCSharp()

using LibVLCSharp.Shared;
namespace Maui.Views;

DemoView.xaml.cs
public partial class DemoView : ContentPage
{
    // LibVLC and MediaPlayer fields
    private readonly LibVLC _libVLC;
    private readonly LibVLCSharp.Shared.MediaPlayer _mediaPlayer;

    public DemoView()
    {
        InitializeComponent();

        BindingContext = DemoViewModel();

        Core.Initialize();

        // Create LibVLC and MediaPlayer instances
        _libVLC = new LibVLC();
        _mediaPlayer = new LibVLCSharp.Shared.MediaPlayer(_libVLC);

        // Attach the MediaPlayer to the MediaPlayerElement (from XAML)
        mediaPlayerElement.MediaPlayer = _mediaPlayer;
    }

    protected override void OnAppearing()
    {
        base.OnAppearing();

        // Create a media object with your stream URL.
        // Replace the URL with your own stream (for example, an RTSP URL).
        // Better resolution - /Preview_01_main
        var media = new Media(_libVLC, "rtsp://user:pass@192.168.0.1/Preview_01_sub", FromType.FromLocation);
  
        // Force the use of TCP for RTSP (so it works with iOS)
        media.AddOption(":rtsp-tcp");

        // Start playback when the page appears.
        _mediaPlayer.Play(media);
    }

    protected override void OnDisappearing()
    {
        base.OnDisappearing();

        // Stop playback and clean up resources.
        _mediaPlayer?.Stop();
        _mediaPlayer?.Dispose();
        _libVLC?.Dispose();
    }
}

DemoView.xaml
<vlc:MediaPlayerElement x:Name="mediaPlayerElement" HeightRequest="400" WidthRequest="400"  />
```

**References:**
- [LibVLCSharp.MAUI.Sample](https://github.com/videolan/libvlcsharp/blob/3.x/samples/LibVLCSharp.MAUI.Sample/MainPage.xaml)
